jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,461 @@
+				if (!newArgTypes) {
+					argTypesAddr = cachedArgLayouts.get(argLayouts).longValue();
+				}
+				
+				initCifNativeThunkData(argLayoutStrs, retLayoutStr, newArgTypes);
+				
+				/* Cache the address of prep_cif and argTypes after setting up via the out-of-line native code */
+				if (newArgTypes) {
+					cachedArgLayouts.put(argLayouts, Long.valueOf(argTypesAddr));
+				}
+				cachedCifNativeThunkAddr.put(funcDescHashPlusFuncAddr, Long.valueOf(cifNativeThunkAddr));
+			}
+		}
+	}
+	
+	/* The method is ultimately invoked by Clinker on the specific platforms to generate the requested
+	 * method handle to the underlying C function.
+	 */
+	@SuppressWarnings("javadoc")
+	public static MethodHandle getBoundMethodHandle(Addressable downcallAddr, MethodType functionMethodType, FunctionDescriptor funcDesc) {
+		ProgrammableInvoker nativeInvoker = new ProgrammableInvoker(downcallAddr, functionMethodType, funcDesc);
+		try {
+			MethodHandle boundHandle = lookup.bind(nativeInvoker, "runNativeMethod", MethodType.methodType(Object.class, Object[].class));
+			
+			/* Replace the original handle with the specified types of the C function */
+			boundHandle = permuteMH(boundHandle, functionMethodType);
+			return boundHandle;
+		} catch (NoSuchMethodException | IllegalAccessException | NullPointerException | WrongMethodTypeException e) {
+			throw new InternalError(e);
+		}
+	}
+	
+	/* Collect and convert the passed-in arguments to an Object array for the underlying native call */
+	private static MethodHandle permuteMH(MethodHandle targetHandle, MethodType nativeMethodType) throws NullPointerException, WrongMethodTypeException {
+		MethodHandle resultHandle = targetHandle.asCollector(0, Object[].class, nativeMethodType.parameterCount());
+		resultHandle = resultHandle.asType(nativeMethodType);
+		return resultHandle;
+	}
+	
+	/* The method (bound by the method handle to the native code) intends to invoke the C function via the inlined code */
+	@SuppressWarnings("boxing")
+	Object runNativeMethod(Object[] args) {
+		int argCount = args.length;
+		long newArgs[] = new long[argCount];
+		Class<?>[] argTypeClasses = funcMethodType.parameterArray();
+		for (int argIndex = 0; argIndex < argCount; argIndex++) {
+			newArgs[argIndex] = convertArgToLongValue(argTypeClasses[argIndex], args[argIndex]);
+		}
+		long returnVal = invokeNative(functionAddr.address().toRawLongValue(), cifNativeThunkAddr, newArgs);
+		
+		/* Process the return value depending on the return type */
+		Object processedReturnVal = processReturnVal(returnVal);
+		return processedReturnVal;
+	}
+	
+	/* Covert arguments to long to adapt to the underlying code dealing with different types
+	 * on various platforms with little/big-endianness.
+	 */
+	@SuppressWarnings("boxing")
+	private static long convertArgToLongValue(Class<?> argTypeClass, Object argValue) {
+		Class<?> realArgTypeClass = unboxingPrimitiveClassType(argTypeClass);
```

Comment:
> Do the generated functionMethodType (in getBoundMethodHandle) always contain the boxed version of primitives? or can they contain both box and unboxed? Or do the boxed versions show up because of the asType() conversion in permuteMH? 

Nope, it won't occur in normal cases as this is not documented in the Spec (and no support from RI as such) but we can provide the users with the flexibility to choose in case that they occasionally pass over to `getBoundMethodHandle()` with the boxed types in which case we are able to handle silently by default.  If there is any updated jtreg test suite that rejects these parameters in the future, we might need to disable/remove that to match the behaviour of RI. So we don't have to intentionally optimize handle the boxed version at this point.

Reply:
Okay, I think it would be good to avoid the extra allocations on the native call path. 

If you change the signature of `runNativeMethod` to `long[] args` then over here you can use:
``` 
MethodHandle filterArguments​(MethodHandle target,
 int pos,
 MethodHandle... filters)
```

Where the filter is essentially the contents of `convertArgToLongValue`.  Eg.

```
if (param.type == float.class) {
     filter = Float::floatToIntBits;
} else if (param.type == double.class) {
    ...
```

For things like `((Character)argValue).charValue();` these can be handled with:

```
} else if (param.type == char.class) {
    filter = ProgrammableInvoker::c2l
} else ...


static final long c2l(char c) {
     return (long) c; //i2l bytecode
}
```

Reply:
with this approach we should also cache the MH for conversion so we are not generating anonClasses on the call path

Reply:
It makes more sense to call `filterArguments​`. Will replace the code involved to see how it goes this way.

Reply:
Updated with `filterArguments​` in `permuteMH()` which works good in the existing test cases